logstash-core/lib/logstash/config/pipeline_config.rb

```diff
@@ -44,5 +44,19 @@ def display_debug_information
       logger.debug("Merged config")
       logger.debug("\n\n#{config_string}")
     end
+
+    def lookup_source_and_line(merged_config_line)
+      remaining_lines = merged_config_line
```

Comment:
iterating over all of the parts for each lookup, doing all of the work to count lines of each config part each time, seems like it would have performance side-effects.

I _think_ we can do the bulk of the work just once, by caching a mapping of file names, offsets, and sizes, and then using that mapping at lookup time instead of going all the way to the config parts.

Below is untested, and may be subject to off-by-one errors 

~~~ ruby

    def lookup_source_and_line(merged_line_number)
      source_map.each do |part_offset, part_lines, part_id|
        rebased_line_number = merged_line_number - part_offset
        next if rebased_line_number > part_lines

        return [part_id, rebased_line_number]
      end
      
      raise IndexError
    end

    private

    def source_map
      @source_map ||= begin
        offset = 0
        source_map = []
        config_parts.each do |config_part|
          part_lines = config_part.getLinesCount()
          @source_map << [offset, part_lines, config_part.id]
          offset += part_lines
        end
        source_map.freeze
      end
    end
~~~